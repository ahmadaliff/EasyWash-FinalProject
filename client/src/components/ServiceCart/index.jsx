import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { FormattedNumber } from 'react-intl';

import { Add, Delete, DryCleaning, Remove } from '@mui/icons-material';
import { IconButton } from '@mui/material';

import { actionChangeQuantity, actionDeleteFromCart } from '@pages/Cart/action';

import classes from '@components/ServiceCart/style.module.scss';

const ServiceCart = ({ service }) => {
  const dispatch = useDispatch();
  return (
    <div className={classes.service} data-testid="service-cart">
      <DryCleaning />
      <div className={classes.serviceInfoWrap}>
        <div className={classes.serviceInfo}>
          <p>{service?.name}</p>
          <p>
            <FormattedNumber
              // eslint-disable-next-line no-unsafe-optional-chaining
              value={service?.price * service?.Carts[0].quantity}
              // eslint-disable-next-line react/style-prop-object
              style="currency"
              currency="IDR"
              minimumFractionDigits={0}
            />
          </p>
        </div>
        <div className={classes.quantity}>
          <div>
            <IconButton
              onClick={() => dispatch(actionChangeQuantity(service?.id, parseInt(service?.Carts[0].quantity, 10) - 1))}
              disabled={parseInt(service?.Carts[0].quantity, 10) - 1 === 0}
              data-testid="button-min"
            >
              <Remove />
            </IconButton>
            {service?.Carts[0].quantity}
            <IconButton
              onClick={() => dispatch(actionChangeQuantity(service?.id, parseInt(service?.Carts[0].quantity, 10) + 1))}
              data-testid="button-plus"
            >
              <Add />
            </IconButton>
          </div>
          <IconButton
            className={classes.delete}
            onClick={() => dispatch(actionDeleteFromCart(service?.Carts[0].id))}
            data-testid="button-delete"
          >
            <Delete />
          </IconButton>
        </div>
      </div>
    </div>
  );
};

ServiceCart.propTypes = {
  service: PropTypes.object,
};

export default ServiceCart;
