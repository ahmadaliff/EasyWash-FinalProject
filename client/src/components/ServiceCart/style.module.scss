@import '@styles/base';
.service {
  @include directionX;
  align-items: center;
  hr {
    margin: 0;
  }
  gap: 0.125rem;
  padding: 0.25rem;
  .serviceInfoWrap {
    width: 100%;
    p {
      margin: 0;
      font-size: small;
    }
    .quantity {
      @include directionX;
      align-items: center;
      justify-content: space-between;
      font-size: medium;
      gap: 0.5rem;
      svg {
        height: 1rem;
        width: 1rem;
      }
    }
    .delete {
      svg {
        height: 1rem;
        width: 1rem;
      }
    }
    .serviceInfo {
      @include directionX;
      justify-content: space-between;
    }
  }
}
