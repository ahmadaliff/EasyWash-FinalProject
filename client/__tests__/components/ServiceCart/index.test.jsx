import ServiceCart from '@components/ServiceCart';
import { fireEvent, render } from '@utils/testHelper';

let wrapper;
const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

const mockProps = {
  service: {
    name: 'cuci kiloan',
    price: 2000,
    Carts: [{ quantity: 1, id: 1 }],
  },
};
beforeEach(() => {
  wrapper = render(<ServiceCart {...mockProps} />);
  jest.clearAllMocks();
});
describe('NoData Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('service-cart')).toBeInTheDocument();
  });
  test('Button min clicked', () => {
    const { getByTestId, getByText } = wrapper;
    const button = getByTestId('button-min');
    fireEvent.click(button);
    expect(getByText('1')).toBeInTheDocument();
  });
  test('Button min clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-plus');
    fireEvent.click(button);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });
  test('Button min clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-delete');
    fireEvent.click(button);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });
  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
