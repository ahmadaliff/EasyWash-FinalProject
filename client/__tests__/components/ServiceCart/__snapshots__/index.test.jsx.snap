// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoData Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="service"
        data-testid="service-cart"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DryCleaningIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.56 11.36 13 8.44V7c0-.55-.45-1-1-1s-1-.45-1-1 .45-1 1-1 1 .45 1 1h2c0-1.84-1.66-3.3-3.56-2.95-1.18.22-2.15 1.17-2.38 2.35-.3 1.56.6 2.94 1.94 3.42v.63l-6.56 2.92c-.88.38-1.44 1.25-1.44 2.2v.01C3 14.92 4.08 16 5.42 16H7v6h10v-6h1.58c1.34 0 2.42-1.08 2.42-2.42v-.01c0-.95-.56-1.82-1.44-2.21zM18.58 14H17v-1H7v1H5.42c-.23 0-.42-.19-.42-.43 0-.17.1-.32.25-.38l6.75-3 6.75 3c.15.07.25.22.25.39 0 .23-.19.42-.42.42z"
          />
        </svg>
        <div
          class="serviceInfoWrap"
        >
          <div
            class="serviceInfo"
          >
            <p>
              cuci kiloan
            </p>
            <p>
              Rp 2.000
            </p>
          </div>
          <div
            class="quantity"
          >
            <div>
              <button
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                data-testid="button-min"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="RemoveIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13H5v-2h14v2z"
                  />
                </svg>
              </button>
              1
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                data-testid="button-plus"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium delete css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="button-delete"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="service"
      data-testid="service-cart"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DryCleaningIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.56 11.36 13 8.44V7c0-.55-.45-1-1-1s-1-.45-1-1 .45-1 1-1 1 .45 1 1h2c0-1.84-1.66-3.3-3.56-2.95-1.18.22-2.15 1.17-2.38 2.35-.3 1.56.6 2.94 1.94 3.42v.63l-6.56 2.92c-.88.38-1.44 1.25-1.44 2.2v.01C3 14.92 4.08 16 5.42 16H7v6h10v-6h1.58c1.34 0 2.42-1.08 2.42-2.42v-.01c0-.95-.56-1.82-1.44-2.21zM18.58 14H17v-1H7v1H5.42c-.23 0-.42-.19-.42-.43 0-.17.1-.32.25-.38l6.75-3 6.75 3c.15.07.25.22.25.39 0 .23-.19.42-.42.42z"
        />
      </svg>
      <div
        class="serviceInfoWrap"
      >
        <div
          class="serviceInfo"
        >
          <p>
            cuci kiloan
          </p>
          <p>
            Rp 2.000
          </p>
        </div>
        <div
          class="quantity"
        >
          <div>
            <button
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="button-min"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="RemoveIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13H5v-2h14v2z"
                />
              </svg>
            </button>
            1
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="button-plus"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium delete css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="button-delete"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
